### DISCUSSION TITLE
Блокировка введения дублированного значения

### DISCUSSION THREAD

**Discussion by juramebel** [totumonline/totum-ru-issues-and-discussions#53] [Как это сделать?] [answered]
Здравствуйте!
Начал делать небольшой блок складского учета и расчета потребностей производства для снабжения.
Все завязано на том, что в одной заявке не могут быть две одинаковые позиции. Потому что при добавлении или изменении создается одна связанная запись в общей таблице потребностей (код ищет запись в ней, удаляет и вставляет новую).
Озадачился таким вопросом.
Поскольку наименование позиций в этих таблицах, это селект, то уникальность ввода данных не отслеживается настройками.
Возможно есть какой то код, который заблокирует введение дубля?

![image](https://user-images.githubusercontent.com/19141501/150408363-ff85dbb6-d27c-4679-b76b-853f59715bc9.png)

**Comment by erpic-ru** [comment, marked as answer]
Можно разместить код действия при добавлении/изменении. Который будет искать в текущей таблице значение текущего поля и выдавать ошибку при совпадениях. Примерно, так:

```
= :if(condition: $find!=''; then: $err; )

find:select(table: $#ntn;  field: 'id'; where: 'pozition' = $#nfv; where: 'id' != #id;)

err: errorException(text: "Такое уже есть в списке")
```

**Reply by juramebel** [reply]
Когда видишь, кажется очевидным) Спасибо)

**Comment by ttm-alexeypolunin** [comment]
Для дополнительного удобства можно не давать выбрать в Селекте повторяющееся значение. Пример дальше написан с учетом того, что добавленные поля изменять нельзя (для оптимизации нагрузки).

Код селекта: 

```
// По пустому id определяем это строка добавления или нет

= : if(condition: #id != ""; then: $srlOne; else: $srlMinus)

// Если это не строка добавления то для показа берем только введенное значение
srlOne: selectRowListForSelect(table: 'table'; field: 'field'; where: 'id' = $#nfv)

// Если строка добавления
srlMinus: selectRowListForSelect(table: 'table'; field: 'field'; where: 'id' = $listminus)

		// Вычитаем и общего списка добавленные в эту таблицу
		listminus: listMinus(list: $listid; list: $listadd)
			listadd: selectList(table: $#ntn; field: $#nf)
				listid: selectList(table: 'table'; field: 'id')
				
```

**Нужно включить Индивидуальный расчет селекта. Годится только для небольших таблиц, например для Расчетный в цикле**

**Reply by juramebel** [reply]
Это отличный вариант определять поведение селекта, я так понимаю.
Например, когда он определяет стадии, сделки к примеру. 
И можно переключать их в определенной последовательности только.

**Reply by ttm-alexeypolunin** [reply]
### Важное замечание по безопасности

Да, важное замечание: по умолчанию значения рассчитанные в поле Селект не являются единственно-допустимыми. Те если чел знает java-script он сможет записать значение отсутствующее в выпадающем списке.

Что бы проверять на сервере введенное значение автоматически надо включить [Проверять введенные значения на сервере](https://docs.totum.online/fields#fields-settings-checkSelectValues).

Тогда машина при сохранении еще раз вычислит список селекта и сравнит введенное значение с ним перед записью в БД. Это небольшой оверхед, но это обязательно к использованию в ответственных местах!
